modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/poll/PollingSourceWrapper.java

```diff
@@ -603,14 +607,19 @@ private void setCurrentWatermarkAsMinimumRejectWatermark(Serializable minimumRej
     watermarkObjectStore.store(WATERMARK_ITEM_OS_KEY, minimumRejectedByLimitPassingWatermark);
   }
 
-  private void updateRecentlyProcessedIds() throws ObjectStoreException {
+  private void updateRecentlyProcessedIds(boolean clearRecentlyProcessed) throws ObjectStoreException {
     Lock osClearingLock = lockFactory.createLock(UPDATE_PROCESSED_LOCK);
     try {
       osClearingLock.lock();
       List<String> strings = idsOnUpdatedWatermark.allKeys();
-      recentlyProcessedIds.clear();
+      if (clearRecentlyProcessed) {
+        recentlyProcessedIds.clear();
+      }
       strings.forEach(key -> {
         try {
+          if (!clearRecentlyProcessed && recentlyProcessedIds.contains(key)) {
```

Comment:
We can try to retrieve all the values of `recentlyProcessedIds ` and then call contains for each key. In this way, we avoid making a request for each key to OSv2.
Not in this PR, it's something we should think about.